independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/VariableProducesSetDefaultContentTypeResponseHandler.java

```diff
@@ -0,0 +1,49 @@
+package org.jboss.resteasy.reactive.server.handlers;
+
+import java.util.List;
+
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.MediaType;
+
+import org.jboss.resteasy.reactive.common.util.MediaTypeHelper;
+import org.jboss.resteasy.reactive.common.util.ServerMediaType;
+import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
+import org.jboss.resteasy.reactive.server.spi.ServerRestHandler;
+
+/**
+ * Handler that defines the default content type when a Response is returned.
+ * While it might not be the final content type, we still need to make sure
+ * the default content type is provided to {@code ContainerResponseFilter}.
+ * <p>
+ * This particular one negotiates the content type when there are multiple ones defined.
+ */
+public class VariableProducesSetDefaultContentTypeResponseHandler implements ServerRestHandler {
+
+    private final ServerMediaType mediaTypeList;
+
+    public VariableProducesSetDefaultContentTypeResponseHandler(ServerMediaType mediaTypeList) {
+        this.mediaTypeList = mediaTypeList;
+    }
+
+    @Override
+    public void handle(ResteasyReactiveRequestContext requestContext) throws Exception {
+        MediaType res = null;
+        List<String> accepts = requestContext.getHttpHeaders().getRequestHeader(HttpHeaders.ACCEPT);
+        for (String accept : accepts) {
+            res = mediaTypeList.negotiateProduces(accept).getKey();
+            if (res != null) {
+                break;
+            }
+        }
+        if (res == null) { // fallback to ensure that MessageBodyWriter is passed the proper media type
+            res = mediaTypeList.negotiateProduces(requestContext.serverRequest().getRequestHeader(HttpHeaders.ACCEPT))
```

Comment:
Oh, I see, I suppose that `requestContext.getHttpHeaders()` might have been overridden by a filter, while `requestContext.serverRequest()` is immutable?
Otherwise I don't understand why these would be different. Besides the fact that I suspect that here, `getRequestHeader` will not properly split the header…
I'm a bit lost here :-/

Reply:
I haven't invented this code, it comes straight from `VariableProducesHandler`. I just copied the exact same behavior as I supposed that's what we wanted.

And I agree I don't really understand what is going on here either.

Reply:
Code comes from this PR: https://github.com/quarkusio/quarkus/pull/40022 .

/cc @geoand

Reply:
Yeah, lovely stuff... Anything media type related is horrible - both in the spec and the impl

Reply:
In any case, I think being consistent in how we handle things for `Response`/not `Response` is a good thing.

Reply:
Definitely agreed

Reply:
Maybe I should add a comment in both classes though so that they are kept in sync if updates are made.

I will do that once we agree with the big picture.

Reply:
A comment would be nice because if we don't understand why it's like this now, we will also not understand it later if an issue is raised and we need to know that there are two places where the same behaviour occurs. Or you refactor the two into a single method, also with a comment saying this is fishy and point at this PR for this discussion, for later archeology.